### ISSUE TITLE
[QUESTION/POSSIBLE USER ERROR]: AXI-Stream VVC byte_endianess not working?

### ISSUE THREAD

**Issue by sibeov** [UVVM/UVVM#196]
# Explanation and question
How does the `byte_endianness` work with regards to the `t_axistream_bfm_config` and its associated VVCs?
It seems to have no effect on the output and input of the VVCs whenever I change it?
I have tried both `FIRST_BYTE_LEFT` and `FIRST_BYTE_RIGHT` (the others as well actually), but there is no change in byte order in Modelsims wave window?
It's probably a user-error on my part or something I have misunderstood.

## Some configurations
32-bit bus interface but I use `t_slv_array`'s with octet width in and out of the VVCs.
```vhdl
  constant my_axistream_bfm_config : t_axistream_bfm_config := (
    max_wait_cycles                => integer'high,
    max_wait_cycles_severity       => ERROR,
    clock_period                   => -1 ns,
    clock_period_margin            => 0 ns,
    clock_margin_severity          => TB_ERROR,
    setup_time                     => -1 ns,
    hold_time                      => -1 ns,
    bfm_sync                       => SYNC_ON_CLOCK_ONLY,
    match_strictness               => MATCH_EXACT,
    byte_endianness                => FIRST_BYTE_LEFT,
    valid_low_at_word_num          => 0,
    valid_low_multiple_random_prob => 0.5,
    valid_low_duration             => 0,
    valid_low_max_random_duration  => 5,
    check_packet_length            => false,
    protocol_error_severity        => ERROR,
    ready_low_at_word_num          => 0,
    ready_low_multiple_random_prob => 0.5,
    ready_low_duration             => 0,
    ready_low_max_random_duration  => 5,
    ready_default_value            => '0',
    id_for_bfm                     => ID_BFM
  );

i_rx_vvc : entity bitvis_vip_axistream.axistream_vvc
    generic map (
      -- When true: This VVC is an AXI4 Stream master. Data is output from BFM.
      -- When false: This VVC is an AXI4 Stream slave. Data is input to BFM.
      gc_vvc_is_master => false,
      gc_data_width    => 32,
      gc_user_width    => 1,
      -- (Note: STRB_WIDTH = DATA_WIDTH/8)
      gc_id_width             => 1,
      gc_dest_width           => 1,
      gc_instance_idx         => 1
      gc_axistream_bfm_config => my_axistream_bfm_config
    )
    port map (
      clk              => csi_th_clk,
      axistream_vvc_if => axis_vvc_rx_if
    );
```

# How to reproduce
Instantiate a "tx" and "rx" VVC, connect to a DUT (can be just a FIFO or a dummy black box transmitting the input to the output) and try all `byte_endianness` possibilities.

# Solution for now
Implemented a "swap" function which swaps the bytes to desired byte order.

**Comment by mariuselv** [comment]
Hi,

Thank you for your message. We appreciate your report and our team will investigate the issue promptly.

Best regards,
Marius

**Comment by erick166** [comment]
Hi,

The config **byte_endianness** in AXI-Stream has 2 possible values: **LOWER_BYTE_LEFT** and **LOWER_BYTE_RIGHT** (_FIRST_BYTE_LEFT_ and _FIRST_BYTE_RIGHT_ are the same and will be removed, while _LOWER_WORD_LEFT_ and _LOWER_WORD_RIGHT_ are not used here).

The purpose of this config is to change the order of the bytes inside the array elements of a t_slv_array, e.g.

```
variable v_data_array_4_byte  : t_slv_array(0 to 1)(31 downto 0);

v_data_array_4_byte := (x"11223344", x"AABBCCDD");
axistream_transmit(AXISTREAM_VVCT, C_MASTER_IDX, v_data_array_4_byte, "4 bytes transmit");
```

With LOWER_BYTE_LEFT config, the order will be:

```
# UVVM: ID_PACKET_DATA                   125.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"11", byte# 0. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   125.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"22", byte# 1. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   125.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"33", byte# 2. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   125.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"44", byte# 3. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   142.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"AA", byte# 4. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   142.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"BB", byte# 5. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   142.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"CC", byte# 6. '4 bytes transmit'  [10]
# UVVM: ID_PACKET_DATA                   142.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"DD", byte# 7. '4 bytes transmit'  [10]
```

Whereas with LOWER_BYTE_RIGHT, the order will be:

```
# UVVM: ID_PACKET_DATA                   175.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"44", byte# 0. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   175.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"33", byte# 1. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   175.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"22", byte# 2. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   175.0 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"11", byte# 3. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   192.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"DD", byte# 4. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   192.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"CC", byte# 5. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   192.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"BB", byte# 6. '4 bytes transmit'  [16]
# UVVM: ID_PACKET_DATA                   192.5 ns  AXISTREAM_VVC,2                axistream_transmit(8 words[8b])=> Tx x"AA", byte# 7. '4 bytes transmit'  [16]
```

Is this how you intended to use it?

**Comment by sibeov** [comment]
Hi.

Thanks for the response and sorry for late response. A lot going on.
Will follow up and check out suggestions.

Kind Regards

**Comment by erick166** [comment]
Hi, any updates on this issue?

**Comment by sibeov** [comment]
> Hi, any updates on this issue?

Hi. Have not had time. Sorry. Will check it out this week.

**Comment by sibeov** [comment]
I 

> # UVVM: ID_PACKET_DATA                   192.5 ns  A

Yes. This was the intention. But I did not seem to change the output as I would have expected. I have not been able to check it out any further since I mentioned the issue. But I was able to test it in the latest version and it worked as expected. I do not mind closing this issue. If anything I can reopen it at a later stage.

Sorry for the very late reply.

Kind Regards
